Issues on using schemathesis with pytest-aiohttp

Hi,

First, thanks for a great library and an effort behind! 

I'm happy to try it, but ran into the issue, after tried to integrate it with my [aiohttp.web](https://aiohttp.readthedocs.io/en/stable/web.html) application via next test case:

```py
from pathlib import Path

import schemathesis
from petstore.app import create_app


schema = schemathesis.from_path(
    Path(__file__).parent / "petstore-expanded.yaml"
)


@schema.parametrize()
async def test_petstore(aiohttp_client, case):
    client = await aiohttp_client(create_app())
    response = await client.request(
        case.method, case.formatted_path, headers=case.headers, json=case.json
    )
    assert response.status < 500
```

Where, 

- `aiohttp_client` is a fixture from [pytest-aiohttp](https://aiohttp.readthedocs.io/en/stable/testing.html#pytest)
- `petstore.app:create_app` factory for creating `web.Application`

This test case results in,

```
E       fixture 'case' not found
>       available fixtures: aiohttp_client, aiohttp_raw_server, aiohttp_server, aiohttp_unused_port, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, cov, doctest_namespace, fast, loop, loop_debug, monkeypatch, no_cover, proactor_loop, pytestconfig, raw_test_server, record_property, record_testsuite_property, record_xml_attribute, recwarn, subtests, test_client, test_server, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_port
>       use 'pytest --fixtures [testpath]' for help on them.
```

after running [pytest](https://pytest.org/en/latest/) as,

```
python -m pytest /path/to/file/with_test.py
```

More data that might be helpful,

- Python 3.7.4
- OS: macOS 10.14.4
- pytest: 5.2.1
- hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/playpauseandstop/Projects/rororo/.hypothesis/examples')
- pytest plugins: schemathesis-0.8.1, subtests-0.2.1, aiohttp-0.3.0, hypothesis-4.39.0, cov-2.8.1

----

Maybe, I got the concept behind `schemathesis` wrong, but I thought that I'm being able to,

- Make the `schema` instance from the OpenAPI 3.0 schema file, which stores next to test case module
- Wrap my test case with `@schema.parametrize()` decorator
- Create my `web.Application` and test `aiohttp_client` inside the test
- Run the test
- And ensure that schema from the file covered well by my aiohttp.web application

If this not possible, I will happy to know on proposed workflow to test my aiohttp.web application with `schemathesis`.

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -11,6 +11,7 @@ Added
 
 - Test executor collects results of execution. `#29`_
 - CLI option ``--base-url`` for specifying base URL of API. `#118`_
+- Support for coroutine-based tests. `#121`_
 
 Changed
 ~~~~~~~
@@ -203,6 +204,7 @@ Fixed
 .. _0.3.0: https://github.com/kiwicom/schemathesis/compare/v0.2.0...v0.3.0
 .. _0.2.0: https://github.com/kiwicom/schemathesis/compare/v0.1.0...v0.2.0
 
+.. _#121: https://github.com/kiwicom/schemathesis/issues/121
 .. _#118: https://github.com/kiwicom/schemathesis/issues/118
 .. _#99: https://github.com/kiwicom/schemathesis/issues/99
 .. _#98: https://github.com/kiwicom/schemathesis/issues/98

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -40,6 +40,7 @@ coverage = "^4.5"
 pytest = ">4.6.4"
 aiohttp = "^3.6"
 pytest-mock = "^1.11.0"
+pytest-asyncio = "^0.10.0"
 
 [tool.poetry.plugins]
 pytest11 = {schemathesis = "schemathesis.extra.pytest_plugin"}

--- a/src/schemathesis/_hypothesis.py
+++ b/src/schemathesis/_hypothesis.py
@@ -1,5 +1,6 @@
 """Provide strategies for given endpoint(s) definition."""
-from typing import Callable, Generator, Optional
+import asyncio
+from typing import Any, Callable, Generator, Optional
 
 import hypothesis
 import hypothesis.strategies as st
@@ -15,9 +16,30 @@ def create_test(endpoint: Endpoint, test: Callable, settings: Optional[hypothesi
     """Create a Hypothesis test."""
     strategy = endpoint.as_strategy()
     wrapped_test = hypothesis.given(case=strategy)(test)
+    original_test = get_original_test(test)
+    if asyncio.iscoroutinefunction(original_test):
+        wrapped_test.hypothesis.inner_test = make_async_test(original_test)  # type: ignore
     if settings is not None:
         wrapped_test = settings(wrapped_test)
     return add_examples(wrapped_test, endpoint)
+
+
+def get_original_test(test: Callable) -> Callable:
+    """Get the original test function even if it is wrapped by `hypothesis.settings` decorator."""
+    if getattr(test, "_hypothesis_internal_settings_applied", False):
+        # `settings` decorator is applied
+        return test._hypothesis_internal_test_function_without_warning  # type: ignore
+    return test
+
+
+def make_async_test(test: Callable) -> Callable:
+    def async_run(*args: Any, **kwargs: Any) -> None:
+        loop = asyncio.get_event_loop()
+        coro = test(*args, **kwargs)
+        future = asyncio.ensure_future(coro, loop=loop)
+        loop.run_until_complete(future)
+
+    return async_run
 
 
 def get_examples(endpoint: Endpoint) -> Generator[Case, None, None]:

--- a/src/schemathesis/extra/pytest_plugin.py
+++ b/src/schemathesis/extra/pytest_plugin.py
@@ -11,11 +11,14 @@ from ..models import Endpoint
 from ..utils import is_schemathesis_test
 
 
+@hookimpl(hookwrapper=True)  # type: ignore
 def pytest_pycollect_makeitem(collector: nodes.Collector, name: str, obj: Any) -> Optional["SchemathesisCase"]:
     """Switch to a different collector if the test is parametrized marked by schemathesis."""
+    outcome = yield
     if is_schemathesis_test(obj):
-        return SchemathesisCase(obj, name, collector)
-    return None
+        outcome.force_result(SchemathesisCase(obj, name, collector))
+    else:
+        outcome.get_result()
 
 
 class SchemathesisCase(PyCollector):

--- a/test/conftest.py
+++ b/test/conftest.py
@@ -53,7 +53,7 @@ def testcmd(testdir):
 
 @pytest.fixture()
 def testdir(testdir):
-    def maker(content, method=None, endpoint=None, **kwargs):
+    def maker(content, method=None, endpoint=None, pytest_plugins=("aiohttp.pytest_plugin",), **kwargs):
         schema = make_schema(**kwargs)
         preparation = dedent(
             """
@@ -70,11 +70,12 @@ def testdir(testdir):
         testdir.makepyfile(preparation, content)
         testdir.makepyfile(
             conftest=dedent(
-                """
+                f"""
+        pytest_plugins = {pytest_plugins}
         def pytest_configure(config):
             config.HYPOTHESIS_CASES = 0
         def pytest_unconfigure(config):
-            print(f"Hypothesis calls: {config.HYPOTHESIS_CASES}")
+            print(f"Hypothesis calls: {{config.HYPOTHESIS_CASES}}")
         """
             )
         )

--- a/test/test_async.py
+++ b/test/test_async.py
@@ -1,0 +1,95 @@
+import pytest
+
+from .utils import integer
+
+
+@pytest.fixture(params=["aiohttp.pytest_plugin", "pytest_asyncio"])
+def plugin(request):
+    return request.param
+
+
+def test_simple(testdir, plugin):
+    # When the wrapped test is a coroutine function and pytest-aiohttp/asyncio plugin is used
+    testdir.make_test(
+        f"""
+async def func():
+    return 1
+
+@schema.parametrize()
+{"@pytest.mark.asyncio" if plugin == "pytest_asyncio" else ""}
+async def test_(request, case):
+    request.config.HYPOTHESIS_CASES += 1
+    assert case.path == "/v1/users"
+    assert case.method == "GET"
+    assert await func() == 1
+""",
+        pytest_plugins=[plugin],
+    )
+    result = testdir.runpytest("-v")
+    result.assert_outcomes(passed=1)
+    # Then it should be executed as any other test
+    result.stdout.re_match_lines([r"test_simple.py::test_\[GET:/v1/users\] PASSED", r"Hypothesis calls: 1"])
+
+
+def test_settings_first(testdir, plugin):
+    # When `hypothesis.settings` decorator is applied to a coroutine-based test before `parametrize`
+    parameters = {"parameters": [integer(name="id", required=True)]}
+    testdir.make_test(
+        f"""
+@schema.parametrize()
+{"@pytest.mark.asyncio" if plugin == "pytest_asyncio" else ""}
+@settings(max_examples=5)
+async def test_(request, case):
+    request.config.HYPOTHESIS_CASES += 1
+    assert case.path == "/v1/users"
+    assert case.method in ("GET", "POST")
+""",
+        pytest_plugins=[plugin],
+        paths={"/users": {"get": parameters, "post": parameters}},
+    )
+    result = testdir.runpytest("-v", "-s")
+    result.assert_outcomes(passed=2)
+    # Then it should be executed as any other test
+    result.stdout.re_match_lines([r"Hypothesis calls: 10$"])
+
+
+def test_aiohttp_client(testdir):
+    # When a wrapped test uses `aiohttp_client` fixture from `aiohttp`
+    testdir.make_test(
+        """
+from aiohttp import web
+import yaml
+
+@pytest.fixture()
+def app():
+    saved_requests = []
+
+    async def schema(request):
+        content = yaml.dump(raw_schema)
+        return web.Response(body=content)
+
+    async def users(request):
+        saved_requests.append(request)
+        return web.Response()
+
+    app = web.Application()
+    app.add_routes([web.get("/swagger.yaml", schema), web.get("/v1/users", users)])
+    app["saved_requests"] = saved_requests
+    return app
+
+@schema.parametrize()
+@settings(max_examples=5)
+async def test_(request, aiohttp_client, app, case):
+    request.config.HYPOTHESIS_CASES += 1
+    client = await aiohttp_client(app)
+    response = await client.request(case.method, case.formatted_path, headers=case.headers, json=case.body)
+    assert response.status < 500
+    assert len(app["saved_requests"]) == 1
+    assert app["saved_requests"][0].method == "GET"
+    assert app["saved_requests"][0].path == "/v1/users"
+"""
+    )
+    result = testdir.runpytest("-v", "-s")
+    result.assert_outcomes(passed=1)
+    # Then it should be executed as any other test
+    result.stdout.re_match_lines([r"Hypothesis calls: 1$"])

--- a/tox.ini
+++ b/tox.ini
@@ -6,6 +6,7 @@ envlist = pylint,mypy,py{36,37,38},coverage-report
 deps =
   aiohttp
   coverage
+  pytest-asyncio
   pytest-mock
 commands =
   coverage run --source=schemathesis -m pytest {posargs:} --junitxml=reports/xunit-tests.xml test

